Transient message boxes in Emacs

david-christiansen opened the pull request:
# Transient message boxes in Emacs

Adds a message display hook to the Emacs mode, and uses it to display boxes containing the results of various commands in the buffer, in addition to the Flycheck interface. The categories of output to display are customizable, and `C-c C-b` toggles the display of boxes in `lean-mode`.

![screenshot from 2017-06-13 17-49-08](https://user-images.githubusercontent.com/115330/27110822-c66b3040-5060-11e7-9830-585c5a3dbdfc.png)

david-christiansen committed:
feat(emacs): Add a message hook to the Emacs mode

When the server protocol returns a message, call user-defined hooks
with that message. This enables custom display of messages in addition
to Flycheck.

david-christiansen committed:
feat(emacs): Add the display of Lean messages in transient boxes

Informational output is usually seen in Flycheck in lean-mode. Now, it
can also be displayed in transient boxes that are not really part of
the buffer. This allows them to be quickly visually compared.

jonsterling commented:
Whoa, this is super sweet. Maybe I'll resurrect RedPRL's emacs mode and try this out...

david-christiansen commented:
@jonsterling The code as written is only somewhat portable to other proof systems, but it could be abstracted into an elisp library if you want it for RedPRL.

Kha commented:
This is nice! Should there be a `(lean-server-show-messages)` on toggling so that the display is changed immediately?
What do you think of using a more efficient presentation for single-line messages, like what IntelliJ does during debugging?
![inline variables view](https://www.jetbrains.com/help/img/idea/2017.1/ij_inline_debugging.png)

david-christiansen commented:
> Should there be a (lean-server-show-messages) on toggling so that the display is changed immediately?

Yes, I think so. Good idea. I can update the PR later today.

> What do you think of using a more efficient presentation for single-line messages, like what IntelliJ does during debugging?

Also a good idea. The only downside that I can think of is that the presentation of the box version and the single-line version would become quite visually different, but that's more of an argument for finding an alternative graphical presentation for the multi-line version. Perhaps instead of a box, it should just be a vertical line on the left combined with putting it in a different face.

david-christiansen committed:
feat(emacs): Immediately toggle display of messages in Emacs

Now, message display is toggled immediately rather than waiting for a
keypress.

david-christiansen committed:
feat(emacs): Make display of message boxes more compact

Message boxes are now displayed without caption (which is available in
a tooltip). Additionally, single-line messages are displayed to the
right of the line that caused them, instead of underneath.

david-christiansen commented:
Now it looks like this:
![screenshot from 2017-06-14 08-04-27](https://user-images.githubusercontent.com/115330/27139614-551b7724-50d8-11e7-8c54-f0ff6c3e310d.png)

Kha commented:
Very nice!

david-christiansen commented:
Thanks!
